plugin/openassetio_manager_bal/BasicAssetLibraryInterface.py

```diff
@@ -352,6 +361,40 @@ def register(
             except Exception as exc:  # pylint: disable=broad-except
                 self.__handle_exception(exc, idx, errorCallback)
 
+    def __validate_publish_entity_traits(self, entity_info, trait_ids, idx, error_callback):
+        if not bal.exists(entity_info, self.__library):
+            # Can publish with any traits if the entity doesn't exist.
+            return True
+
+        entity = bal.entity(entity_info, self.__library)
+        # Entities must be published with the same trait set as they
+        # currently have (can be overridden per access mode).
+        if not set(entity.traits.keys()).issubset(trait_ids):
+            error_callback(
+                idx,
+                BatchElementError(
+                    BatchElementError.ErrorCode.kInvalidTraitSet,
+                    "Publishing to this entity requires traits that are missing from the input",
+                ),
+            )
+            return False
+        return True
+
+    def __validate_publish_policy(self, traits_data, access, idx, error_callback):
```

Comment:
[Question] A little confused about the split between validate policy and validate entity traits, as policy is also concerned with specific traits, given that the management policy makes decision based on a trait set. Are we making the same decision twice here?

Reply:
It's a good question. It comes back to the overlap between `entityTraits` (for `kWrite`) and `managementPolicy` (for `kWrite`/`kRequired`).

For validation purposes, `managementPolicy` only tells us whether a trait set is publishable _at all_ (i.e. if `ManagedTrait` is imbued). It can further tell us which traits are persistable, but its too late to use that knowledge when we're already in a publishing method - i.e. hosts are free to publish traits that cannot be persisted, it's not an error, just a no-op.

`entityTraits` for `kWrite`, on the other hand, tells us which traits _must_ be provided in order to publish to a particular entity. So we can compare that against the trait set provided to the publishing method, and validate the provided trait set is a superset of the required trait set.